src/interpreter/balances.rs

```diff
@@ -0,0 +1,323 @@
+            );
+
+            ofs + AssetId::LEN + WORD_SIZE
+        });
+}
+
+#[test]
+fn try_from_iter_wont_overflow() {
+    use crate::prelude::*;
+    use rand::rngs::StdRng;
+    use rand::{Rng, SeedableRng};
+
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+
+    let a: AssetId = rng.gen();
+    let b: AssetId = rng.gen();
+    let c: AssetId = rng.gen();
+
+    // Sanity check
+    let balances = vec![(a, u64::MAX), (b, 15), (c, 0)];
+    let runtime_balances = RuntimeBalances::try_from_iter(balances.clone()).expect("failed to create balance set");
+
+    balances.iter().for_each(|(asset, val)| {
+        let bal = runtime_balances.balance(asset).expect("failed to fetch balance");
+
+        assert_eq!(val, &bal);
+    });
+
+    // Aggregated sum check
+    let balances = vec![(a, u64::MAX), (b, 15), (c, 0), (b, 1)];
+    let balances_aggregated = vec![(a, u64::MAX), (b, 16), (c, 0)];
+    let runtime_balances = RuntimeBalances::try_from_iter(balances.clone()).expect("failed to create balance set");
+
+    balances_aggregated.iter().for_each(|(asset, val)| {
+        let bal = runtime_balances.balance(asset).expect("failed to fetch balance");
+
+        assert_eq!(val, &bal);
+    });
+
+    // Overflow won't panic
+    let balances = vec![(a, u64::MAX), (b, 15), (c, 0), (a, 1)];
+    let err = RuntimeBalances::try_from_iter(balances.clone()).expect_err("overflow set should fail");
+
+    assert_eq!(ValidationError::ArithmeticOverflow, err);
+}
+
+#[test]
+fn checked_add_and_sub_works() {
+    use crate::prelude::*;
+    use rand::rngs::StdRng;
+    use rand::{Rng, SeedableRng};
+
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+
+    let mut memory = Interpreter::without_storage().memory;
+
+    let asset: AssetId = rng.gen();
+
+    let balances = vec![(asset, 0)];
+    let mut balances = RuntimeBalances::try_from_iter(balances).expect("failed to create set");
+
+    // Sanity check
+    let bal = balances.balance(&asset).expect("failed to fetch balance");
+    assert_eq!(bal, 0);
+
+    // Add zero balance not in the set should result in zero and not mutate the set
+    let asset_b: AssetId = rng.gen();
+    assert_ne!(asset, asset_b);
+
+    let val = balances
+        .checked_balance_add(&mut memory, &asset_b, 0)
+        .expect("failed to add balance");
+
+    assert_eq!(val, 0);
+    assert!(balances.balance(&asset_b).is_none());
+
+    // Normal add balance works
+    let val = balances
+        .checked_balance_add(&mut memory, &asset, 150)
+        .expect("failed to add balance");
+    let bal = balances.balance(&asset).expect("failed to fetch balance");
+
+    assert_eq!(val, 150);
+    assert_eq!(bal, 150);
+
+    let val = balances
+        .checked_balance_add(&mut memory, &asset, 75)
+        .expect("failed to add balance");
+    let bal = balances.balance(&asset).expect("failed to fetch balance");
+
+    assert_eq!(val, 225);
+    assert_eq!(bal, 225);
```

Comment:
We should also test that the call to `_set_memory_balance` is working as expected. Currently, we only verify the VM memory access for initialization, although the VM memory is also modified whenever these balances change.